src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Push down projections into various operators. TODO.
+
+use std::collections::{HashMap, HashSet};
+
+use expr::{RelationExpr, ScalarExpr};
+use repr::{RelationType, Row};
+
+use crate::TransformArgs;
+
+/// Push down projections TODO
+#[derive(Debug)]
+pub struct ProjectionPushdown;
+
+impl crate::Transform for ProjectionPushdown {
+    fn transform(
+        &self,
+        relation: &mut RelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        self.action(relation, (0..relation.typ().column_types.len()).collect());
+        Ok(())
+    }
+}
+
+impl ProjectionPushdown {
+    /// Attempt to push a projection as far down the tree as possible. Logically the outcome of
+    /// this function should be that relation will be replaced with relation.project(projection),
+    /// but we will attempt to push things down further if we can.  Further, some operators (say,
+    /// Reduce) may be able to synthesize additional projections that we will attempt to push down.
+    pub fn action(&self, relation: &mut RelationExpr, mut projection: Vec<usize>) {
+        match relation {
+            RelationExpr::Project { input, outputs } => {
+                self.action(input, projection.iter().map(|c| outputs[*c]).collect());
+                *relation = input.take_dangerous();
+            }
+            RelationExpr::Union { base, inputs } => {
+                for inp in inputs {
+                    self.action(inp, projection.clone());
+                }
+                self.action(base, projection);
+            }
+            RelationExpr::Constant { rows, typ } => {
+                *relation = RelationExpr::Constant {
+                    rows: rows
+                        .iter()
+                        .map(|(row, m)| {
+                            let datums = row.unpack();
+                            (Row::pack(projection.iter().map(|i| datums[*i])), *m)
+                        })
+                        .collect(),
+                    typ: RelationType::new(
+                        projection
+                            .iter()
+                            .map(|i| typ.column_types[*i].clone())
+                            .collect(),
+                    ),
+                };
+            }
+            RelationExpr::Filter { input, predicates } => {
+                let input_arity = input.arity();
+                let mut needed = HashSet::new();
```

Comment:
Maybe minor, maybe helpful, but this could plausibly be structured as appending in to `projection`, sorting, deduping, and then using `.binary_search` to do the column rewrites. Only if bouncing between hash sets, hash maps, and all that is distracting.